File: backend/app/modules/fund_nav/schemas/history.py
from __future__ import annotations

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field, model_validator


class FundNavHistorySyncRequest(BaseModel):
    mode: Literal["recent_days", "date_range"] = "recent_days"
    recent_days: int | None = Field(default=30, ge=1, le=3650)
    start_date: date | None = None
    end_date: date | None = None
    workers: int = Field(default=4, ge=1, le=16)

    @model_validator(mode="after")
    def validate_range(self) -> "FundNavHistorySyncRequest":
        if self.mode == "recent_days":
            if self.recent_days is None:
                raise ValueError("recent_days is required")
            return self
        if self.start_date is None or self.end_date is None:
            raise ValueError("start_date and end_date are required")
        if self.start_date > self.end_date:
            raise ValueError("start_date must not be after end_date")
        return self


class FundNavHistorySyncStartOut(BaseModel):
    task_id: int | None = None
    pid: int
    started: bool
    start_date: str
    end_date: str
    workers: int
    stdout_log: str
    stderr_log: str
    message: str


class FundNavHistoryProgressCount(BaseModel):
    status: str
    count: int


class FundNavHistoryProgressItem(BaseModel):
    fund_code: str
    fund_name: str | None = None
    status: str
    started_at: datetime | None = None
    finished_at: datetime | None = None
    duration_seconds: float | None = None
    error: str | None = None


class FundNavHistorySyncStatusOut(BaseModel):
    running: bool
    task_id: int | None = None
    pid: int | None = None
    start_date: str
    end_date: str
    workers: int | None = None
    stdout_log: str | None = None
    stderr_log: str | None = None
    counts: list[FundNavHistoryProgressCount]
    latest_done: list[FundNavHistoryProgressItem]
    running_items: list[FundNavHistoryProgressItem]
    failed_items: list[FundNavHistoryProgressItem]


class FundNavHistoryTaskOut(BaseModel):
    id: int
    task_type: str
    status: str
    start_date: str
    end_date: str
    workers: int
    total_count: int
    success_count: int
    failed_count: int
    running_count: int
    skipped_count: int
    retry_count: int
    pid: int | None = None
    stdout_log: str | None = None
    stderr_log: str | None = None
    message: str | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None
    duration_seconds: float | None = None
    created_at: datetime


class FundNavHistoryTaskListOut(BaseModel):
    tasks: list[FundNavHistoryTaskOut]


class FundNavHistoryTaskDetailOut(FundNavHistoryTaskOut):
    counts: list[FundNavHistoryProgressCount]
    latest_done: list[FundNavHistoryProgressItem]
    running_items: list[FundNavHistoryProgressItem]
    failed_items: list[FundNavHistoryProgressItem]

File: frontend/src/modules/fund_nav/api/history.ts
import { apiClient } from '../../../api/client'

export type FundNavHistorySyncMode = 'recent_days' | 'date_range'

export interface FundNavHistorySyncRequest {
  mode: FundNavHistorySyncMode
  recent_days?: number | null
  start_date?: string | null
  end_date?: string | null
  workers: number
}

export interface FundNavHistorySyncStartResult {
  task_id?: number | null
  pid: number
  started: boolean
  start_date: string
  end_date: string
  workers: number
  stdout_log: string
  stderr_log: string
  message: string
}

export interface FundNavHistoryProgressCount {
  status: string
  count: number
}

export interface FundNavHistoryProgressItem {
  fund_code: string
  fund_name?: string | null
  status: string
  started_at?: string | null
  finished_at?: string | null
  duration_seconds?: number | null
  error?: string | null
}

export interface FundNavHistorySyncStatus {
  running: boolean
  task_id?: number | null
  pid?: number | null
  start_date: string
  end_date: string
  workers?: number | null
  stdout_log?: string | null
  stderr_log?: string | null
  counts: FundNavHistoryProgressCount[]
  latest_done: FundNavHistoryProgressItem[]
  running_items: FundNavHistoryProgressItem[]
  failed_items: FundNavHistoryProgressItem[]
}

export interface FundNavHistoryTask {
  id: number
  task_type: string
  status: string
  start_date: string
  end_date: string
  workers: number
  total_count: number
  success_count: number
  failed_count: number
  running_count: number
  skipped_count: number
  retry_count: number
  pid?: number | null
  stdout_log?: string | null
  stderr_log?: string | null
  message?: string | null
  started_at?: string | null
  finished_at?: string | null
  duration_seconds?: number | null
  created_at: string
}

export interface FundNavHistoryTaskDetail extends FundNavHistoryTask {
  counts: FundNavHistoryProgressCount[]
  latest_done: FundNavHistoryProgressItem[]
  running_items: FundNavHistoryProgressItem[]
  failed_items: FundNavHistoryProgressItem[]
}

export async function startFundNavHistorySync(payload: FundNavHistorySyncRequest): Promise<FundNavHistorySyncStartResult> {
  const { data } = await apiClient.post<FundNavHistorySyncStartResult>('/fund-nav/history-sync/start', payload, {
    timeout: 30000,
  })
  return data
}

export async function stopFundNavHistorySync(): Promise<{ stopped: boolean; message: string; task_id?: number | null; pid?: number | null }> {
  const { data } = await apiClient.post<{ stopped: boolean; message: string; task_id?: number | null; pid?: number | null }>(
    '/fund-nav/history-sync/stop',
  )
  return data
}

export async function getFundNavHistorySyncStatus(): Promise<FundNavHistorySyncStatus> {
  const { data } = await apiClient.get<FundNavHistorySyncStatus>('/fund-nav/history-sync/status')
  return data
}

export async function listFundNavHistorySyncTasks(): Promise<FundNavHistoryTask[]> {
  const { data } = await apiClient.get<{ tasks: FundNavHistoryTask[] }>('/fund-nav/history-sync/tasks')
  return data.tasks
}

export async function getFundNavHistorySyncTask(taskId: number): Promise<FundNavHistoryTaskDetail> {
  const { data } = await apiClient.get<FundNavHistoryTaskDetail>(`/fund-nav/history-sync/tasks/${taskId}`)
  return data
}

export async function rerunFundNavHistorySyncTask(taskId: number): Promise<FundNavHistorySyncStartResult> {
  const { data } = await apiClient.post<FundNavHistorySyncStartResult>(`/fund-nav/history-sync/tasks/${taskId}/rerun`)
  return data
}
